category_encoders/target_encoder.py

```diff
@@ -113,19 +120,33 @@ def __init__(self, verbose=0, cols=None, drop_invariant=False, return_df=True, h
                           category=FutureWarning)
         self.mapping = None
         self._mean = None
-        if hierarchy:
+        if isinstance(hierarchy, (dict, pd.DataFrame)) and cols is None:
+            raise ValueError('Hierarchy is defined but no columns are named for encoding')
```

Comment:
why do we need this error? the current behaviour is that if you do not specify `cols` all categorical columns will be determined and encoded. So if i rely on this default and not specify columns this will break now right?

Reply:
ah ok, in line 138 you have to know at init time what the columns will look like, determining them on the fly during fit-time will not be enough

Reply:
in the postcode example cols should be just `postcode` right? so specifying `cols=None` would not make any sense anyway since the HIER_postcode_N cols would be encoded as columns on their own instead of being recognized in the hierarchy.  
Then this is just fine to throw an error here